### ISSUE TITLE
Possible memory leak in `SET_STRUCT_FIELD`

### ISSUE THREAD

**Issue by lausdahl** [overturetool/vdm2c#29]
It looks like there might be a memory leak in relation to `SET_STRUCT_FIELD` this clones the new value and sets the local struct field with the new value. However it does not free the previous value so the pointer could be lost.

Note that `SET_FIELD` uses be method `SET_STRUCT_FIELD`.

``` c
/*
 * Macro to set the (sub-)class specific field from a class struct
 */
#define SET_STRUCT_FIELD(tname,ptr,fieldtype,fieldname,newValue) (*( (fieldtype*) (  ((unsigned char*)ptr) + offsetof(struct tname, fieldname) )  )=vdmClone(newValue))
```

When this is seen in combination with `GET_FIELD` which always clones the return value. The there is not way to free a value that has been set.

``` c
/*
 * Macro to obtain a field from a (sub-)class specific class struct. We clone to preserve value semantics and the rule of freeing
 */
#define GET_FIELD(thisTypeName,fieldTypeName,classValue,fieldName) vdmClone(GET_STRUCT_FIELD(fieldTypeName,CLASS_CAST(TO_CLASS_PTR(classValue,thisTypeName),thisTypeName,fieldTypeName) ,struct TypedValue*,m_##fieldTypeName##_##fieldName))
```

**Comment by bandurvp** [comment]
Hi Kenneth, I have just pushed a change where the old value is checked and freed if it is not NULL before the new value is set at that field.  But I think there is more to it.  If that previous value needs reference semantics, then it should not be freed, I think?  Also, I do not understand the purpose of the last part of the comment, regarding GET_FIELD.

**Comment by lausdahl** [comment]
Reference semantics: It should also be freed since it hold a reference to the object and decrements its counter once freed. Then when the last reference is freed when the object is too.

if `GET_FIELD` returns a clone and `SET_FIELD` sets a new. Then there is no way to free previous values that were set. Thats what the last part says. If you fixed `SET_FIELD` so it cleans up then it should be ok. I assume that you call just call `vdmFree` on the field before setting it. I think that should do since a `NULL` value is ignored by free and that any other is freed (and class values decrements the count of frees it all)

**Comment by bandurvp** [comment]
I checked for NULL in the macro, but you're right, vdmFree does this itself, I'll change that.  Should I close the issue then?

**Comment by lausdahl** [comment]
yes if it works close it with your fix so there is a reference to the commit

**Comment by bandurvp** [comment]
Is that done with "Cose issue #29"?

**Comment by lausdahl** [comment]
no with

close
closes
closed
fix
fixes
fixed
resolve
resolves
resolved
followed by #29

so fixed #29 will do it

see https://help.github.com/articles/closing-issues-via-commit-messages/ https://help.github.com/articles/closing-issues-via-commit-messages/

**Comment by lausdahl** [comment]
ok but that does not work. It is:

close #29
